Cross-platform SCI-F

bilke opened the pull request:
A follow-up on #69. Provides `scif()` (for installing the `scif` executable) and `scif_app()` (for installing a Scientific Filesystem) building blocks.

## Example usage:

In recipe:

```python
Stage0 += scif()
Stage0 += scif_app(
    name='mpi-bw',
    run='mpi-bandwidth',
    install=[
        'wget -q -nc --no-check-certificate -P src '
        'https://computing.llnl.gov/tutorials/mpi/samples/C/mpi_bandwidth.c',
        'mkdir -p bin',
        'mpicc -o bin/mpi-bandwidth src/mpi_bandwidth.c'
    ],
    help='This app provides a MPI bandwidth test program',
    test='mpirun -np 2 mpi-bandwidth'
)
```

Generate, build and run:

```sh
hpccm --recipe mpi_bw_scif_recipe.py --format Docker --out _out
docker build -t mpi_bw_scif -f _out/Dockerfile .
docker run mpi_bw_scif scif run mpi-bw
```

Example is not that ingenious as the bandwith executable should of course be wrapped by `mpirun` but you get the idea.

This also allows to convert to Singularity and run:

```sh
docker run -v /var/run/docker.sock:/var/run/docker.sock \
  -v $PWD:/output --privileged -t --rm \
  singularityware/docker2singularity --name mpi_bw_scif.simg \
  mpi_bw_scif
singularity run --app mpi-bw mpi_bw_scif.simg
# OR
singularity run mpi_bw_scif.simg scif run mpi-bw
```

Which is really helpful during development of container definitions with hpccm as you can use Dockers image layer caching for developing Singularity container.

samcmill commented:
What is the overall vision of SCIF and HPCCM?  There is the existing `_app` options in the primitives that are Singularity specific but well integrated with the HPCCM workflow.  This here is more general in that it works with Docker and Singularity, but the SCIF "manifest" is a separate specification block and not well integrated with the rest of HPCCM.

(BTW: The workflow of building with Docker to take advantage of the layer caching and then converting the image to Singularity is one that I use very often.)

After building with Docker and converting to Singularity, is this expected?
```
$ singularity test --app mpi-bw mpi_bw_scif.simg 
No Singularity tests for contained app: mpi-bw
```

```
$ singularity run --app mpi-bw mpi_bw_scif.simg 
No Singularity runscript for contained app: mpi-bw
```

bilke commented:
I really like the approach of SCIF providing an interface to several apps in a container. I also like the integration with HPCCM with the `_app` option.

I dislike the proposed SCIF implementation here with the separate manifest file but furthermore really like the workflow building for Docker and then converting to Singularity.. So this PR is for getting the best of both worlds, i.e. Singularity container apps and Docker layer caching but the tradeoff is a not so elegant integration into HPCCM...

No, I expected that the `--app`-syntax should work. See https://github.com/vsoch/scif/issues/51. I will investigate this further.

samcmill commented:
If you squint, SCIF looks like "containers-inside-containers".  With that perspective, here's a sketch of what a more HPCCM-like approach to SCIF might look like.  I'm sure there are important implementation details that I'm glossing over.  The key idea is to introduce an analogue of `Stage` for SCIF "containers".  Would love to get feedback on this type of approach.

```
Stage0 += baseimage(...)

s = scif(names=['mpi-bw', '...'])
Stage0 += s

# SCIF manifest.  Created automatically as a side effect of the scif() building block.  
s.manifest[0] += shell(commands=[...]) # Used as %appinstall
s.manifest[0] += runscript(commands=[...]) # Used as %apprun
s.manifest[0] += comment('...') # Used as %apphelp
s.manifest[0] += environment(variables={...}) # Used as %appenv
s.manifest[0] += copy(...) # Used as %appfiles
s.manifest[0] += label(...) # Used as %applabels
s.manifest[0] += ??? # Used as %apptest.  New primitive type?

s.manifest[1] += ...

# Perhaps this syntax would be preferable?
# s.manifest['mpi-bw'] += ...

# Magic!  For Docker, this would generate the SCIF manifest files, the COPY instructions and
# the "RUN scif install ..." steps.  For Singularity, this would generate %appinstall, %apprun, etc. 
# blocks.
```

A variant on the above, where there isn't the automatic side effect of creating a manifest list when using the `scif()` building block.

```
Stage0 += baseimage(...)

# Explicitly create a list of Scif manifests, modeled on how Stages are created.
s = [Scif(name='mpi-bw'), Scif(name='...'), ...] 

Stage0 += scif(manifests=s)

# SCIF manifest
s[0] += shell(commands=[...]) # Used as %appinstall
s[0] += runscript(commands=[...]) # Used as %apprun
s[0] += comment('...') # Used as %apphelp
s[0] += environment(variables={...}) # Used as %appenv
s[0] += copy(...) # Used as %appfiles
s[0] += label(...) # Used as %applabels
s[0] += ??? # Used as %apptest.  New primitive type?

s[1] += ...

# Magic!  For Docker, this would generate the SCIF manifest files, the COPY instructions and
# the "RUN scif install ..." steps.  For Singularity, this would generate %appinstall, %apprun, etc. 
# blocks.
```

samcmill commented:
BTW: I don't know if "manifest" is proper terminology for the corresponding SCIF concept or not.  The correct term should be used, whatever it is.

bilke commented:
@samcmill I have refactored the scif-building block as you suggested. For the user it became even easier. There is no need to create an explicit list of scif recipes; just create a `scif()`, `+=` add layers to it and add it to a stage that's it. Uses Singularity's native syntax and the `scif`-tool on Docker. I will also add `%apphelp` and `%apptest` primitives.

samcmill commented on hpccm/building_blocks/scif.py:
```diff
@@ -0,0 +1,105 @@
+                packages(ospackages=['python-pip', 'python-setuptools']),
+                shell(commands=[
+                  'pip install wheel',
+                  'pip install scif==0.0.76'
```
make the version user configurable?  default to 0.0.76 if not specified.

samcmill commented on RECIPES.md:
```diff
@@ -1556,6 +1556,45 @@ Stage1 += p.runtime()
+Parameters:
+
+- `name`: The name of the scif app. **Required**
+- `entrypoint`: Boolean flag to specify wheter this scif app should also be the
```
typo: wheter -> whether

bilke commented on hpccm/recipe.py:
```diff
@@ -68,10 +68,11 @@
 from hpccm.building_blocks.pgi import pgi
 from hpccm.building_blocks.pnetcdf import pnetcdf
 from hpccm.building_blocks.python import python
+from hpccm.building_blocks.scif import scif
```
`__name__` does not change when imported with `as`. I googled a bit but did not found a solution besides class inheritance..

samcmill replied:
too bad...

bilke commented:
@samcmill I addressed your points. When the current state is ok for you then I would also finish the documentation with the experimental label you proposed.

Yep, using building blocks in a scif app would be a super cool feature...

samcmill commented on hpccm/building_blocks/scif.py:
```diff
@@ -0,0 +1,122 @@
+                  'pip install wheel',
+                  'pip install scif=={0}'.format(self.scif_version)
+                ]),
+                comment("End SCI-F installation")
```
FYI: A pip building block was recently added.  Not sure if it would be appropriate to use here or not, so just making sure you're aware.

One approach would be to remove the pip stuff from the scif building block and just say in the documentation that you need to call `pip(packages=['scif'])` before using the scif building block.  There is precedence for this with the MPI and OFED building blocks or some of the IO building blocks and HDF5, for example.

On the other hand, it is convenient to have scif installed automatically as part of the scif building block.   It's your call.

samcmill commented on hpccm/building_blocks/scif.py:
```diff
@@ -0,0 +1,122 @@
+            logging.exception(
+                'Duplicate `{0}` primitive given for scif app. '
+                'Each primitive is allowed just once!'.format(class_name))
+            raise()  # TODO: the logging exception is not raised?!
```
Not a big deal, but the exception stack traces that get printed out are a little ugly.  Use `logging.error()` / `raise` as used elsewhere?

Also a TODO.

samcmill commented on hpccm/building_blocks/scif.py:
```diff
@@ -0,0 +1,105 @@
+        """Install the runtime from a full build in a previous stage.  In this
+           case there is no difference between the runtime and the
+           full build."""
+        return str(self)
```
With #110, this TODO has become more important.

bilke commented:
Closed in favor of #136.
